File: server/routes/reports.py
from flask import Blueprint, render_template, jsonify, request
from flask_login import login_required
from models import Employee, Activity, Screenshot
from extensions import db
from datetime import datetime, timedelta

reports_bp = Blueprint('reports', __name__)

@reports_bp.route('/reports')
@login_required
def reports_page():
    employees = Employee.query.all()
    return render_template('reports.html', employees=employees)

@reports_bp.route('/api/reports/activity/<int:emp_id>')
@login_required
def employee_activity_report(emp_id):
    days = int(request.args.get('days', 7))
    start_date = datetime.utcnow() - timedelta(days=days)

    activities = Activity.query.filter(
        Activity.employee_id == emp_id,
        Activity.recorded_at >= start_date
    ).order_by(Activity.recorded_at.asc()).all()

    daily_data = {}
    for act in activities:
        day = act.recorded_at.strftime('%Y-%m-%d')
        if day not in daily_data:
            daily_data[day] = {'clicks': 0, 'keys': 0, 'idle_minutes': 0, 'active_minutes': 0}
        daily_data[day]['clicks'] += act.mouse_clicks
        daily_data[day]['keys'] += act.keystrokes
        daily_data[day]['idle_minutes'] += act.idle_time // 60
        if act.idle_time < 300:
            daily_data[day]['active_minutes'] += 1

    labels = sorted(daily_data.keys())
    clicks = [daily_data[d]['clicks'] for d in labels]
    keys = [daily_data[d]['keys'] for d in labels]
    idle = [daily_data[d]['idle_minutes'] for d in labels]
    active = [daily_data[d]['active_minutes'] for d in labels]

    return jsonify({
        'labels': labels,
        'clicks': clicks,
        'keystrokes': keys,
        'idle_minutes': idle,
        'active_minutes': active
    })

@reports_bp.route('/api/reports/screenshots/<int:emp_id>')
@login_required
def employee_screenshot_count(emp_id):
    days = int(request.args.get('days', 7))
    start_date = datetime.utcnow() - timedelta(days=days)

    screenshots = Screenshot.query.filter(
        Screenshot.employee_id == emp_id,
        Screenshot.captured_at >= start_date
    ).all()

    daily = {}
    for s in screenshots:
        day = s.captured_at.strftime('%Y-%m-%d')
        daily[day] = daily.get(day, 0) + 1

    labels = sorted(daily.keys())
    counts = [daily[d] for d in labels]

    return jsonify({'labels': labels, 'counts': counts})

@reports_bp.route('/api/reports/summary')
@login_required
def summary_report():
    days = int(request.args.get('days', 7))
    start_date = datetime.utcnow() - timedelta(days=days)
    employees = Employee.query.all()

    summary = []
    for emp in employees:
        act_count = Activity.query.filter(
            Activity.employee_id == emp.id,
            Activity.recorded_at >= start_date
        ).count()
        ss_count = Screenshot.query.filter(
            Screenshot.employee_id == emp.id,
            Screenshot.captured_at >= start_date
        ).count()
        latest = Activity.query.filter_by(employee_id=emp.id).order_by(Activity.recorded_at.desc()).first()

        total_clicks = db.session.query(db.func.sum(Activity.mouse_clicks)).filter(
            Activity.employee_id == emp.id,
            Activity.recorded_at >= start_date
        ).scalar() or 0
        total_keys = db.session.query(db.func.sum(Activity.keystrokes)).filter(
            Activity.employee_id == emp.id,
            Activity.recorded_at >= start_date
        ).scalar() or 0

        summary.append({
            'id': emp.id,
            'name': emp.name,
            'department': emp.department,
            'status': emp.status,
            'activity_count': act_count,
            'screenshot_count': ss_count,
            'total_clicks': total_clicks,
            'total_keys': total_keys,
            'last_active': latest.recorded_at.isoformat() if latest else None
        })

    return jsonify({'summary': summary, 'days': days})

File: node_server/routes/reports.js
const express = require('express');
const router = express.Router();
const { queryAll, queryOne } = require('../db');
const { requireLogin } = require('../middleware/auth');

router.get('/reports', requireLogin, (req, res) => {
  const employees = queryAll('SELECT * FROM employees');
  res.render('reports.html', { employees });
});

router.get('/api/reports/activity/:empId', requireLogin, (req, res) => {
  const days = parseInt(req.query.days) || 7;
  const startDate = new Date(Date.now() - days * 24 * 60 * 60 * 1000).toISOString();

  const activities = queryAll('SELECT * FROM activities WHERE employee_id = ? AND recorded_at >= ? ORDER BY recorded_at ASC', [parseInt(req.params.empId), startDate]);

  const dailyData = {};
  for (const act of activities) {
    const day = act.recorded_at.split(' ')[0];
    if (!dailyData[day]) dailyData[day] = { clicks: 0, keys: 0, idle_minutes: 0, active_minutes: 0 };
    dailyData[day].clicks += act.mouse_clicks;
    dailyData[day].keys += act.keystrokes;
    dailyData[day].idle_minutes += Math.floor(act.idle_time / 60);
    if (act.idle_time < 300) dailyData[day].active_minutes += 1;
  }

  const labels = Object.keys(dailyData).sort();
  res.json({
    labels,
    clicks: labels.map(d => dailyData[d].clicks),
    keystrokes: labels.map(d => dailyData[d].keys),
    idle_minutes: labels.map(d => dailyData[d].idle_minutes),
    active_minutes: labels.map(d => dailyData[d].active_minutes)
  });
});

router.get('/api/reports/screenshots/:empId', requireLogin, (req, res) => {
  const days = parseInt(req.query.days) || 7;
  const startDate = new Date(Date.now() - days * 24 * 60 * 60 * 1000).toISOString();

  const screenshots = queryAll('SELECT * FROM screenshots WHERE employee_id = ? AND captured_at >= ?', [parseInt(req.params.empId), startDate]);

  const daily = {};
  for (const s of screenshots) {
    const day = s.captured_at.split(' ')[0];
    daily[day] = (daily[day] || 0) + 1;
  }

  const labels = Object.keys(daily).sort();
  res.json({ labels, counts: labels.map(d => daily[d]) });
});

router.get('/api/reports/summary', requireLogin, (req, res) => {
  const days = parseInt(req.query.days) || 7;
  const startDate = new Date(Date.now() - days * 24 * 60 * 60 * 1000).toISOString();
  const employees = queryAll('SELECT * FROM employees');

  const summary = employees.map(emp => {
    const actCount = queryOne('SELECT COUNT(*) as count FROM activities WHERE employee_id = ? AND recorded_at >= ?', [emp.id, startDate]).count;
    const ssCount = queryOne('SELECT COUNT(*) as count FROM screenshots WHERE employee_id = ? AND captured_at >= ?', [emp.id, startDate]).count;
    const latest = queryOne('SELECT * FROM activities WHERE employee_id = ? ORDER BY recorded_at DESC LIMIT 1', [emp.id]);
    const totalClicks = queryOne('SELECT COALESCE(SUM(mouse_clicks), 0) as total FROM activities WHERE employee_id = ? AND recorded_at >= ?', [emp.id, startDate]).total;
    const totalKeys = queryOne('SELECT COALESCE(SUM(keystrokes), 0) as total FROM activities WHERE employee_id = ? AND recorded_at >= ?', [emp.id, startDate]).total;

    return {
      id: emp.id, name: emp.name, department: emp.department, status: emp.status,
      activity_count: actCount, screenshot_count: ssCount,
      total_clicks: totalClicks, total_keys: totalKeys,
      last_active: latest ? latest.recorded_at : null
    };
  });

  res.json({ summary, days });
});

module.exports = router;
